openedx_authz/api/data.py

```diff
@@ -0,0 +1,258 @@
+    # So we don't have to manage each subclass separately or hardcoded anywhere.
+
+
+@define
+class ScopeData(AuthZData):
+    """A scope is a context in which roles and permissions are assigned.
+
+    Attributes:
+        scope_id: The scope identifier (e.g., 'org@Demo').
+
+    Acts as a factory: automatically returns the correct subclass based on the scope_id prefix.
+    """
+
+    NAMESPACE: str = "sc"
+    scope_id: str = ""
+    name: str = ""
+
+    def __attrs_post_init__(self):
+        """Ensure scope ID has appropriate namespace prefix."""
+        if not self.scope_id:
+            self.scope_id = f"{self.NAMESPACE}{self.SEPARATOR}{self.name}".lower()
+
+        # Allow reverse lookup of name from scope_id
+        if (
+            not self.name
+            and self.scope_id
+            and self.NAMESPACE
+            and self.scope_id.startswith(f"{self.NAMESPACE}{self.SEPARATOR}")
+        ):
+            self.name = self.scope_id.split(self.SEPARATOR, 1)[1].lower()
+
+
+@define
+class ContentLibraryData(ScopeData):
+    """A content library is a collection of content items.
+
+    Attributes:
+        library_id: The content library identifier (e.g., 'library-v1:edX+DemoX+2021_T1').
+        scope_id: Inherited from ScopeData, auto-generated from library_id if not provided.
+    """
+
+    NAMESPACE: str = "lib"
+    library_id: str = ""
+
+    def __attrs_post_init__(self):
+        """Ensure scope ID has 'lib@' namespace prefix."""
+        if not self.scope_id:
+            self.scope_id = f"{self.NAMESPACE}{self.SEPARATOR}{self.library_id}".lower()
+
+        # Allow reverse lookup of library_id from scope_id
+        if not self.library_id and self.scope_id.startswith(
+            f"{self.NAMESPACE}{self.SEPARATOR}"
+        ):
+            self.library_id = self.scope_id.split(self.SEPARATOR, 1)[1].lower()
+
+
+@define
+class SubjectData(AuthZData):
+    """A subject is an entity that can be assigned roles and permissions.
+
+    Attributes:
+        subject_id: The subject identifier namespaced (e.g., 'user@john_doe').
+
+    Acts as a factory: automatically returns the correct subclass based on the subject_id prefix.
+    """
+
+    NAMESPACE: str = "sub"
+    subject_id: str = ""
+    name: str = ""
+
+    def __attrs_post_init__(self):
+        """Ensure subject ID has appropriate namespace prefix."""
+        if not self.subject_id:
+            self.subject_id = f"{self.NAMESPACE}{self.SEPARATOR}{self.name}".lower()
+
+        # Allow reverse lookup of name from subject_id
+        if not self.name and self.subject_id.startswith(
+            f"{self.NAMESPACE}{self.SEPARATOR}"
+        ):
+            self.name = self.subject_id.split(self.SEPARATOR, 1)[1].lower()
+
+
+@define
+class UserData(SubjectData):
+    """A user is a subject that can be assigned roles and permissions.
+
+    Attributes:
+        username: The username for the user (e.g., 'john_doe').
```

Comment:
When using username in this way, is there a plan to handle username changes, as happens in retirement? I think that without that these will become orphaned or worse, if that usename somehow gets re-used could grant unexpected permissions. A possibility would be to delete all AuthZ rows upon retirement or to use the user id instead.